loader/src/hooks/AddExtraKeys.mm

```diff
@@ -0,0 +1,253 @@
+    }
+    CCKeyboardDispatcher::get()->dispatchKeyboardMSG(keyCode, true, [event isARepeat]);
+}
+
+static void(*s_keyUpExecOrig)(void*, SEL, NSEvent*);
+void keyUpExecHook(void* self, SEL sel, NSEvent* event) {
+    bool extraKey = isExtraKey(event);
+    bool numpad = isKeyNumpad(event);
+    if (!extraKey && !numpad) {
+        return s_keyUpExecOrig(self, sel, event);
+    }
+    if (CCIMEDispatcher::sharedDispatcher()->hasDelegate()) {
+        return s_keyUpExecOrig(self, sel, event);
+    }
+
+    enumKeyCodes keyCode = enumKeyCodes::KEY_Unknown;
+    if (extraKey) {
+        keyCode = extraKeyToKeyCode(event);
+    }
+    if (numpad) {
+        keyCode = numpadToKeyCode(event);
+    }
+    CCKeyboardDispatcher::get()->dispatchKeyboardMSG(keyCode, false, [event isARepeat]);
+}
+
+static void(*s_mouseDownExecOrig)(void*, SEL, NSEvent*);
+void mouseDownExecHook(void* self, SEL sel, NSEvent* event) {
+    if (!isExtraMouseButton(event)) {
+        return s_mouseDownExecOrig(self, sel, event);
+    }
+
+    enumKeyCodes keyCode = mouseButtonToKeyCode(event);
+    CCKeyboardDispatcher::get()->dispatchKeyboardMSG(keyCode, true, false);
+}
+
+static void(*s_mouseUpExecOrig)(void*, SEL, NSEvent*);
+void mouseUpExecHook(void* self, SEL sel, NSEvent* event) {
+    if (!isExtraMouseButton(event)) {
+        return s_mouseUpExecOrig(self, sel, event);
+    }
+
+    enumKeyCodes keyCode = mouseButtonToKeyCode(event);
+    CCKeyboardDispatcher::get()->dispatchKeyboardMSG(keyCode, false, false);
+}
+
+
+class $modify(CCKeyboardDispatcher) {
+    GEODE_FORWARD_COMPAT_DISABLE_HOOKS("CCKeyboardDispatcher new keys")
+
+    const char* keyToString(enumKeyCodes key) {
+        if (key < 0x1000) {
+            return CCKeyboardDispatcher::keyToString(key);
+        }
+
+        switch (key) {
+            case KEY_GraveAccent:
+                return "`";
+            case KEY_OEMEqual:
+                return "=";
+            case KEY_LeftBracket:
+                return "[";
+            case KEY_RightBracket:
+                return "]";
+            case KEY_Backslash:
+                return "\\";
+            case KEY_Semicolon:
+                return ";";
+            case KEY_Apostrophe:
+                return "'";
+            case KEY_Slash:
+                return "/";
+            case KEY_NumEnter:
+                return "NumEnter";
+            case KEY_World1:
+                return "INTL-1";
+            case KEY_World2:
+                return "INTL-2";
+            case MOUSE_4:
+                return "Mouse 4";
+            case MOUSE_5:
+                return "Mouse 5";
+            case MOUSE_6:
+                return "Mouse 6";
+            case MOUSE_7:
+                return "Mouse 7";
+            case MOUSE_8:
+                return "Mouse 8";
+            default:
+                return CCKeyboardDispatcher::keyToString(KEY_Unknown);
+        }
+    }
+};
+
+
+#define OBJC_SWIZZLE(klass, methodName) \
```

Comment:
my new concern with swizzling (that i just made up) is that i think geode doing this would mess with other mods that also swizzle those methods - it might be good to make it a manual hook instead

ideally those mods would also switch to hooks but that requires i do something so... at least we can avoid making the problem worse lol

Reply:
I hook these methods in BetterInputs and it mostly works fine, but sure, I could change to manual hooks instead of swizzling